### ISSUE TITLE
Zip executable

### ISSUE THREAD

**Issue by boxfoot** [git-for-windows/git#775]
- [X] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen)
   or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue
   matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? 32-bit or 64-bit? Include the
   output of `git version` as well.

```
$ git --version
git version 2.7.0.windows.2
```

Using 64-bit version

 - Which version of Windows are you running? 32-bit or 64-bit?

Windows 7 Enterprise SP1 64-bit

 - What options did you set as part of the installation? Or did you choose the
   defaults?

Don't remember - don't think it's relevant

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

Nope

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Bash

- summary

Git for windows ships with a command line `unzip` executable (for me, installed in `/c/Program Files/Git/usr/bin/unzip.exe`), but no zip program.  I'm curious whether it would be possible to include `zip` as well to simplify building cross-platform scripts to support tooling/TechOps for my team, since many of our developers use the bash environment from git for windows as their primary development environment.  

I'm not sure if this is a git for windows issue or upstream, any pointers would be helpful.  Thanks!

**Comment by fourpastmidnight** [comment]
What about `tar`? There's also `bzip2`.

**Comment by boxfoot** [comment]
We have a requirement (because of another system we're using) to use zip.   I know I could use a windows-specific program (like `7za`) but that means our scripts need to be smart enough to identify which systemt they're on and run the right executable....

**Comment by boxfoot** [comment]
But the fact that `gzip` and `bzip2` are both provided is part of why `zip` feels like it could be easily included and isn't.

**Comment by fourpastmidnight** [comment]
It's not that it isn't "easily included", it's just that it's not required by git. This is not a general "linux utilities distribution for Windows", this is a distribution of Git for Windows. As such, all that's included are those things that are required for working with Git in a Windows environment. (If you want a more "general purpose" Linux-y type environment on Windows, you may want to use Cygwin instead.) But more importantly, it sounds like you have everything you need already!

You can call Windows command line utilities from within the BASH (but only if you did not choose the option to only use Git from Bash during installation--if you used the defaults when installing Git for Windows, you'll be fine). For example, type `where -?` at the BASH prompt. That's the Windows `where` program (which is like the `which` Linux utility). So there's nothing stopping you from calling `7z.exe` from the BASH command line (just make sure `7z` is in your `PATH` environment variable, or you'll have to use the fully qualified path to `7z`, which by default in BASH would be `"/c/Program Files (x86)/7-zip/7z.exe"`).

Other than that, if you really want to include a utility such as `zip` in your Git for Windows distribution, you only have two other good options right now:

1. Create a customized installer for Git for Windows based off of the Portable installer (which is basically a 7zip self-extracting archive). This would allow you to specify optional packages to be included with the main Git for Windows distributables.
2. Create a NuGet package for Git for Windows. (This may be able to be installed using chocolatey, but you'll have to find out if this is supported--I really don't know much about chocolatey and what's required for making NuGet packages that work with it). The NuGet packager also allows optional packages to be specified.

You'll need to download and install the Git for Windows SDK to build and package Git for either of the two options above.

Please do not attempt to use the MSI-based installer--it's under heavy development and is not ready for use (e.g. not all the installation options are implemented yet). At this time, the InnoSetup-based installer does not allow extra packages to be specified on the command line.

**Comment by boxfoot** [comment]
That's a totally fair response.  Wasn't looking for git to provide a full linux environment, but thought I'd poke since all the other related binaries seemed to be provided.   I agree that there are lots of other solutions possible, and I'll choose one of them!

**Comment by dscho** [comment]
FWIW Git for Windows is a subset of [MSYS2](https://msys2.github.io/). Maybe you were looking for MSYS2, and not looking hard enough?

**Comment by r2evans** [comment]
I'm familiar with msys2 and I've been using it for a couple of years (initially msys2-direct, then GfW SDK), but its (pacman's) ability to self-corrupt and require a complete reinstall is frustrating (twice in 6 months). Though not unlike Windows itself in this regard, I was grateful that its reinstall was relatively simple and just wasted time and lot of bandwidth. I've really appreciated having the full linuxy toolsets available within the GfW SDK.

I have no perception that I "deserve" or "am entitled" to anything: frankly, I'm very grateful that I can get as much as I do out of the GfW (non-SDK) installation. The fact that it seems intentionally reduced is as much a sign that the maintainers (1) have to deal with a lot of requests for software that "obviously must be included, duh", all the while (2) keeping things small, simple, and maintainable, on top of (3) dealing with venting non-contributing issue-makers.

TL;DR: Though I personally feel that having `zip` in addition to `unzip` should not be adding any complexity (no added libraries or downloads on the maintainers part), neither am I ponying up efforts to contribute. I'm able to replace most of the tools from the SDK that I no longer have: `zip` from my local install of Rtools, and `emacs` from vgoulet's distro. And life goes on. (Wondering if/when bash-for-windows will ever come close to doing what many people have inferred it would.)

Bottom line:

# Thanks for your work!

... it would be nice but I understand you have bigger fish to fry. Cheers.

**Comment by weberjn** [comment]
For many developers Git for Windows has become the de facto Linux subsystem for Windows. It is the best integrated bash for Windows I know, I use it for all command line development work, not only git.
So it would be nice if you could add more Unix tools, e.g. archivers, pwgen

**Comment by fourpastmidnight** [comment]
Thank you for your remarks. The goal of Git for Windows is to provide an up-to-date distribution of Git that works on Windows. Since Git was created by Linus Torvalds (primarily) for Linux kernel development, it required a lot of work to port the tool such that it works well on Windows. This project does a great job of improving the Git core for Windows and Linux users alike. And it's a _lot_ of work.

For this reason, Git for Windows focuses on providing the best distribution for running Git on Windows. Git requires some other Linux utilities, so as a consequence, these utilities are also included in the Git for Windows distribution. And for those who only want to run Git on Windows, this project is an excellent choice.

But some, such as yourself, want/need a bit more. And that's great. What many don't realize is that Git for Windows is just a subset of the MSYS2 project, which itself descends from Cygwin; both of which can provide a complete Unix/Linux/Posix environment on Windows.

When you need more than just Git and its supporting utilities on Windows, one of these two distributions should be your first stop. By keeping the focus of Git for Windows narrow, we aim to be the best implementation of Git running on Windows. If we were to focus on anything more, Git for Windows would not be as good, and the project would be duplicating the efforts of the excellent MSYS2 and Cygwin projects. When you need more than just Git, please give these excellent projects a try, and provide feedback to them, too. They have their hands full as well.

Thank you again for your kind comments.

**Comment by PhilipOakley** [comment]
A third way to access the Msys2 environment is via the Git-for-Windows SDK ([System Development Kit](https://github.com/git-for-windows/build-extra/releases/latest)) which neatly packages all of your favourite Git and bash utilities in the MSYS2 environment with full access to the 'pacman' package manager.

And it allows you to develop, compile and test your own variant of GfW, and hopefully [**contribute**](https://git-for-windows.github.io/#contribute) patches upstream to GfW and/or Git.

The contribute section also has a link to the [SDK technical overview](https://github.com/git-for-windows/git/wiki/Technical-overview)
